### PR TITLE
ENG-12742 micro opt refactor

### PR DESCRIPTION
**Pull request by yzhang1991**

### CODE REVIEW

File: src/frontend/org/voltdb/expressions/AbstractSubqueryExpression.java [hunk 1]
```diff
@@ -89,7 +89,7 @@ public void resetSubqueryNodeId() {
     // Create a matching PVE for this expression to be used on the EE side
     // to get the original expression value
     protected void addCorrelationParameterValueExpression(AbstractExpression expr, List<AbstractExpression> pves) {
-        int paramIdx = AbstractParsedStmt.NEXT_PARAMETER_ID++;
+        int paramIdx = AbstractParsedStmt.NEXT_PARAMETER_ID.getAndIncrement();
```
Comment (wolffcm) [inline src/frontend/org/voltdb/expressions/AbstractSubqueryExpression.java]: Hmm... parameter IDs should all be adjacent for the same statement, right?  If there were concurrent planners that might not be the case and havoc could result.

I think `NEXT_PARAMETER_ID` should be a normal (non-static) member of some object that is not shared by more than one thread.  `QueryPlanner` seems to fit this description?

Comment (wolffcm) [inline src/frontend/org/voltdb/expressions/AbstractSubqueryExpression.java]: Seems like this applies to all of the static variables that you've made Atomic.

File: src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java [hunk 2]
```diff
@@ -58,7 +59,7 @@ protected void recursivelyPrint(AbstractPlanNode node, StringBuilder sb)
     }
```
Comment (wolffcm) [inline src/frontend/org/voltdb/planner/microoptimizations/ReplaceWithIndexLimit.java]: This class has a member variable called `indent` which will incur a data race in `recursivelyPrint`.  This is easily fixable though by just making `indent` a parameter to `recursivelyPrint`.  I realize it's debug-only code.

File: src/frontend/org/voltdb/expressions/AbstractSubqueryExpression.java (hunk 1 above)
Comment (yzhang1991) [inline src/frontend/org/voltdb/expressions/AbstractSubqueryExpression.java]: I don't think the parameter ID has to be adjacent to work, but it makes a lot of sense to see them be adjacent for each statement.

**Pull request merged**